lighthouse-core/gather/gather-runner.js

```diff
@@ -441,91 +486,109 @@ class GatherRunner {
-        baseArtifacts.devtoolsLogs[passConfig.passName] = passData.devtoolsLog;
+  /**
+   * Starting from about:blank, load the page and run gatherers for this pass.
+   * @param {LH.Gatherer.PassContext} passContext
+   * @return {Promise<{artifacts: Partial<LH.GathererArtifacts>, pageLoadError?: LHError}>}
+   */
+  static async runPass(passContext) {
+    /** @type {Partial<GathererResults>} */
+    const gathererResults = {};
+    const {driver, passConfig} = passContext;
 
-        const userAgentEntry = passData.devtoolsLog.find(entry =>
-          entry.method === 'Network.requestWillBeSent' &&
-          !!entry.params.request.headers['User-Agent']
-        );
+    // Go to about:blank, set up, and run `beforePass()` on gatherers.
+    await GatherRunner.loadBlank(driver, passConfig.blankPage);
+    await GatherRunner.setupPassNetwork(passContext);
+    const isPerfPass = GatherRunner.isPerfPass(passContext);
+    if (isPerfPass) await driver.cleanBrowserCaches(); // Clear disk & memory cache if it's a perf run
+    await GatherRunner.beforePass(passContext, gathererResults);
 
-        if (userAgentEntry && !baseArtifacts.NetworkUserAgent) {
-          // @ts-ignore - guaranteed to exist by the find above
-          baseArtifacts.NetworkUserAgent = userAgentEntry.params.request.headers['User-Agent'];
-        }
+    // Navigate, start recording, and run `pass()` on gatherers.
+    await GatherRunner.beginRecording(passContext);
+    await GatherRunner.loadPage(driver, passContext);
+    await GatherRunner.pass(passContext, gathererResults);
+    const loadData = await GatherRunner.endRecording(passContext);
 
-        // If requested by config, save pass's trace.
-        if (passData.trace) {
-          baseArtifacts.traces[passConfig.passName] = passData.trace;
-        }
+    // Disable throttling so the afterPass analysis isn't throttled
+    await driver.setThrottling(passContext.settings, {useThrottling: false});
 
-        if (isFirstPass) {
-          // Copy redirected URL to artifact in the first pass only.
-          baseArtifacts.URL.finalUrl = passContext.url;
-          isFirstPass = false;
-        }
-      }
-      const resetStorage = !options.settings.disableStorageReset;
-      if (resetStorage) await driver.clearDataForOrigin(options.requestedUrl);
-      await GatherRunner.disposeDriver(driver);
-      return GatherRunner.collectArtifacts(gathererResults, baseArtifacts);
-    } catch (err) {
-      // cleanup on error
-      GatherRunner.disposeDriver(driver);
-      throw err;
+    // Save devtoolsLog and trace.
+    const baseArtifacts = passContext.baseArtifacts;
+    baseArtifacts.devtoolsLogs[passConfig.passName] = loadData.devtoolsLog;
+    if (loadData.trace) baseArtifacts.traces[passConfig.passName] = loadData.trace;
+
+    // If there were any load errors, treat all gatherers as if they errored.
+    const pageLoadError = GatherRunner.getPageLoadError(passContext, loadData);
```

Comment:
there are a few ways #8340 could be incorporated here. One would be have `GatherRunner.loadPage` pass back non-fatal errors as it does in that PR and pass that error into `GatherRunner.getPageLoadError` here, and have the code in `getPageLoadError` decide if the error is severe enough, which error wins if there are multiple ones, etc

(where exactly this call should go is another question that we need to figure out in terms of what's reasonable to pass back in an errored run...e.g. it doesn't make much sense to save a trace of a security interstitial)